src/shape.rs

```diff
@@ -660,237 +677,214 @@ impl ShapeLine {
                         } else if wrap == Wrap::Glyph {
                             for (glyph_i, glyph) in word.glyphs.iter().enumerate().rev() {
-                                let glyph_size = font_size * glyph.x_advance;
-                                if fit_x - glyph_size >= 0. {
-                                    fit_x -= glyph_size;
-                                    word_range_width += glyph_size;
+                                let glyph_width = font_size * glyph.x_advance;
+                                if fit_x - glyph_width >= 0. {
+                                    fit_x -= glyph_width;
+                                    word_range_width += glyph_width;
                                     continue;
                                 } else {
-                                    word_ranges.push((
+                                    self.add_to_visual_line(
+                                        &mut current_visual_line,
+                                        span_index,
                                         (i, glyph_i + 1),
                                         fitting_start,
                                         word_range_width,
                                         number_of_blanks,
-                                    ));
+                                    );
+                                    visual_lines.push(current_visual_line);
+                                    current_visual_line = VisualLine::default();
+
                                     number_of_blanks = 0;
-                                    fit_x = line_width - glyph_size;
-                                    word_range_width = glyph_size;
+                                    fit_x = line_width - glyph_width;
+                                    word_range_width = glyph_width;
                                     fitting_start = (i, glyph_i + 1);
                                 }
                             }
                         } else {
                             // Wrap::Word
-                            let mut prev_word_width = None;
-                            if word.blank && number_of_blanks > 0 {
-                                // current word causing a wrap is a space so we ignore it
-                                number_of_blanks -= 1;
-                            } else if let Some(previous_word) = span.words.get(i - 1) {
-                                // Current word causing a wrap is not whitespace, so we ignore the
-                                // previous word if it's a whitespace
-                                if previous_word.blank {
-                                    number_of_blanks -= 1;
-                                    prev_word_width = Some(previous_word.x_advance * font_size);
+                            let mut trialing_space_width = None;
+                            if i > 0 {
+                                if let Some(previous_word) = span.words.get(i - 1) {
```

Comment:
Hmm, this is something I changed to `i + 1` in my PR. I'm just curious, so can you explain to me why this is correct? Or should the variable be renamed to `next_word`?

Reply:
```suggestion
                                if let Some(previous_word) = span.words.get(i + 1) {
```
I still believe this should be `i + 1`, because we are iterating over descending `i` and the last word added to the line had the index `i + 1`, not `i - 1`.

Reply:
Also, I think the `i > 0` check in the incongruent case is wrong. In the congruent case we don't have a previous word if `i == 0`, but this is not true in the incongruent direction. There we don't have a "previous" word if `i + 1 == span.words.len()`, but we don't need to take special care of that because `get` returns `None` if the index is out of bounds.

Reply:
Good catch. It should be `i + 1`.